File: medium/1268. Search Suggestions System/1268. Search Suggestions System.py
from collections import defaultdict
from typing import List


class Trie:
    def __init__(self) -> None:
        self.children = {}
        self.isEnd = False


class Solution:
    def suggestedProducts(
        self, products: List[str], searchWord: str
    ) -> List[List[str]]:
        trie = defaultdict(list)
        products.sort()

        for i in range(1, len(searchWord) + 1):
            pref = searchWord[:i]

            for product in products:
                if len(trie[pref]) == 3:
                    break

                if product.startswith(pref):
                    trie[pref].append(product)

        return [x for x in trie.values()]


solution = Solution()
print(
    solution.suggestedProducts(
        products=["mobile", "mouse", "moneypot", "monitor", "mousepad"],
        searchWord="mouse",
    )
)

File: medium/1268. Search Suggestions System/1268. Search Suggestions System.go
package main

import (
	"sort"
)

func suggestedProducts(products []string, searchWord string) [][]string {
	res := [][]string{}
	sort.Strings(products)
	prev := products

	for i := 0; i < len(searchWord); i++ {
		temp := []string{}
		for _, product := range prev {
			if len(products) > i && product[i] == searchWord[i] {
				temp = append(temp, product)
			}
		}

		prev = temp
		if len(temp) <= 3 {
			res = append(res, temp)
		} else {
			res = append(res, temp[:3])
		}
	}

	return res
}
